test/utility/test_util.i

```diff
@@ -134,3 +134,44 @@ testutil_make_work_dir(char *dir)
 		    MKDIR_COMMAND, dir);
 	free(buffer);
 }
+
+#undef	M_W
+#define	M_W	(rnd)[0]
+#undef	M_Z
+#define	M_Z	(rnd)[1]
+
+/*
+ * testutil_random_init --
+ *	Initialize return of a 32-bit pseudo-random number.
+ */
+static inline void
+testutil_random_init(uint32_t *rnd)
+{
+	M_W = 521288629;
+	M_Z = 362436069;
+}
+
+/*
+ * testutil_random --
+ *	Return a 32-bit pseudo-random number.
+ *
+ * This is an implementation of George Marsaglia's multiply-with-carry pseudo-
+ * random number generator.  Computationally fast, with reasonable randomness
+ * properties.
+ *
+ * We have to be very careful about races here.  Multiple threads can call
+ * __wt_random concurrently, and it is okay if those concurrent calls get the
+ * same return value.  What is *not* okay is if reading the shared state races
+ * with an update and uses two different values for m_w or m_z.  That could
+ * result in a value of zero, in which case they would be stuck on zero
+ * forever.  Take local copies of the shared values to avoid this.
```

Comment:
@markbenvenuto, @michaelcahill:

The hidden assumption here is the two "seeds" the PRNG uses (M_W and M_Z) can never go to 0, if they do, bad stuff happens because they never change after that point. It's worse if the low bits (M_W) go to zero, because then `random % 4` is going to be fixed.

We don't know they don't go to zero eventually anyway, of course, but let's assume that. (I'm currently talking to someone about the period of the function at the moment, so more information soon, I hope.)

Anyway, I still don't understand why this function is thread-safe: if thread 1 reads M_W, and thread 2 writes M_Z, then thread 1 reads M_Z, I don't see why there's not a problem.

It's vanishingly unlikely, and I don't think we currently use `__wt_random` from multiple threads at the same time inside the engine (not sure about the test programs), but I don't understand why the change Michael made in 4a850c8 was sufficient.

As Mark said, we could read/write the two values as 8B quantities and then we'd get atomic behavior.